bindings/rust/bench/src/harness.rs

```diff
@@ -140,49 +140,33 @@ impl CryptoConfig {
     }
 }
 
+/// The TlsBenchConfig trait allows us to map benchmarking parameters to
+/// a configuration object
+pub trait TlsBenchConfig: Sized {
+    /// Make a config with given parameters
+    fn make_config(
+        mode: Mode,
+        crypto_config: CryptoConfig,
+        handshake_type: HandshakeType,
+    ) -> Result<Self, Box<dyn Error>>;
+}
+
+/// The TlsConnection object can be created from a corresponding config type.
+/// Any "purpose specific" config creation should be implemented on the config
+/// rather than the connection to separate configuration vs implementation.
 pub trait TlsConnection: Sized {
     /// Library-specific config struct
     type Config;
```

Comment:
Not 100% sure what you mean by "purpose specific".

Also does it make sense to restrict Config: `type Config: TlsConfig`. Seems like you need that for it to be useful and you would be able to remove the bounds elsewhere.

Reply:
Consider a situation where I want do a self-talk test with s2n-tls security policies. In that case the only thing that I need to configure the connection is the security policy name. But then the current `make_config` method makes no sense.
```
fn make_config(
        mode: Mode,
        crypto_config: CryptoConfig,
        handshake_type: HandshakeType,
    ) -> Result<Self::Config, Box<dyn Error>>;
```
My "security policy self talk" test doesn't need to know about, or care about the crypto config. I have my own idea of how a connection should be configured.

Instead I could now implement a `SecurityPolicySelfTalk` trait on the S2nTls config object, which might just be something like
```
trait SecurityPolicySelfTalk {
    /// return a security policy with appropriately configured default certificates, 
    /// using the specified security policy.
    security_policy_config(sp: &str) -> Self
}
```
Then I can still use the handshake harnesses to actually drive the handshaking.

(this example is a bad one, but hopefully shows what I'm getting at)

Reply:
> Also does it make sense to restrict Config: type Config: TlsConfig. Seems like you need that for it to be useful and you would be able to remove the bounds elsewhere.

Hmm. I'm not following. What is the `TlsConfig` trait that you are referring to? I don't want to require any specific functionality to be implemented on the Config object, because that forces you to implement that functionality on all of s2n-tls, rustls, and openssl. That might not always be logical or feasible.

More generally I want to push bounds down as far I can to avoid making the implementation unnecessarily restrictive.